From f0ad504e8ac167551e9bc41daa904a9fc51a98c9 Mon Sep 17 00:00:00 2001
From: Loc Vu <loc.vu.zn@renesas.com>
Date: Thu, 27 Sep 2018 15:43:25 +0700
Subject: [PATCH 036/433] soc: renesas: rcar-sysc: Add R8A7744 support

Add support for RZ/G1N (R8A7744) SoC power areas to the R-Car SYSC
driver.

Signed-off-by: thongsyho <thong.ho.px@rvc.renesas.com>
Signed-off-by: Loc Vu <loc.vu.zn@renesas.com>
---
 drivers/soc/renesas/Makefile       |  1 +
 drivers/soc/renesas/r8a7744-sysc.c | 33 +++++++++++++++++++++++++++++++++
 drivers/soc/renesas/rcar-sysc.c    |  3 +++
 drivers/soc/renesas/rcar-sysc.h    |  1 +
 4 files changed, 38 insertions(+)
 create mode 100644 drivers/soc/renesas/r8a7744-sysc.c

diff --git a/drivers/soc/renesas/Makefile b/drivers/soc/renesas/Makefile
index 4272a51..818f817 100644
--- a/drivers/soc/renesas/Makefile
+++ b/drivers/soc/renesas/Makefile
@@ -1,5 +1,6 @@
 obj-$(CONFIG_ARCH_RCAR_GEN2)	+= rcar-rst.o
 obj-$(CONFIG_ARCH_R8A7743)	+= rcar-sysc.o r8a7743-sysc.o
+obj-$(CONFIG_ARCH_R8A7744)	+= rcar-sysc.o r8a7744-sysc.o
 obj-$(CONFIG_ARCH_R8A7745)	+= rcar-sysc.o r8a7745-sysc.o
 obj-$(CONFIG_ARCH_R8A77470)	+= rcar-sysc.o r8a77470-sysc.o
 
diff --git a/drivers/soc/renesas/r8a7744-sysc.c b/drivers/soc/renesas/r8a7744-sysc.c
new file mode 100644
index 0000000..e06d62b
--- /dev/null
+++ b/drivers/soc/renesas/r8a7744-sysc.c
@@ -0,0 +1,33 @@
+/*
+ * Renesas R-Car M2-W/N System Controller
+ *
+ * Copyright (C) 2016 Glider bvba
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ */
+
+#include <linux/bug.h>
+#include <linux/kernel.h>
+
+#include <dt-bindings/power/r8a7744-sysc.h>
+
+#include "rcar-sysc.h"
+
+static const struct rcar_sysc_area r8a7744_areas[] __initconst = {
+	{ "always-on",	    0, 0, R8A7744_PD_ALWAYS_ON,	-1, PD_ALWAYS_ON },
+	{ "ca15-scu",	0x180, 0, R8A7744_PD_CA15_SCU,	R8A7744_PD_ALWAYS_ON,
+	  PD_SCU },
+	{ "ca15-cpu0",	 0x40, 0, R8A7744_PD_CA15_CPU0,	R8A7744_PD_CA15_SCU,
+	  PD_CPU_NOCR },
+	{ "ca15-cpu1",	 0x40, 1, R8A7744_PD_CA15_CPU1,	R8A7744_PD_CA15_SCU,
+	  PD_CPU_NOCR },
+	{ "sh-4a",	 0x80, 0, R8A7744_PD_SH_4A,	R8A7744_PD_ALWAYS_ON },
+	{ "sgx",	 0xc0, 0, R8A7744_PD_SGX,	R8A7744_PD_ALWAYS_ON },
+};
+
+const struct rcar_sysc_info r8a7744_sysc_info __initconst = {
+	.areas = r8a7744_areas,
+	.num_areas = ARRAY_SIZE(r8a7744_areas),
+};
diff --git a/drivers/soc/renesas/rcar-sysc.c b/drivers/soc/renesas/rcar-sysc.c
index fb5e3b5..46b8b6f 100644
--- a/drivers/soc/renesas/rcar-sysc.c
+++ b/drivers/soc/renesas/rcar-sysc.c
@@ -278,6 +278,9 @@ static const struct of_device_id rcar_sysc_matches[] = {
 #ifdef CONFIG_ARCH_R8A7743
 	{ .compatible = "renesas,r8a7743-sysc", .data = &r8a7743_sysc_info },
 #endif
+#ifdef CONFIG_ARCH_R8A7744
+	{ .compatible = "renesas,r8a7744-sysc", .data = &r8a7744_sysc_info },
+#endif
 #ifdef CONFIG_ARCH_R8A7745
 	{ .compatible = "renesas,r8a7745-sysc", .data = &r8a7745_sysc_info },
 #endif
diff --git a/drivers/soc/renesas/rcar-sysc.h b/drivers/soc/renesas/rcar-sysc.h
index 7734edd..a340d57 100644
--- a/drivers/soc/renesas/rcar-sysc.h
+++ b/drivers/soc/renesas/rcar-sysc.h
@@ -53,5 +53,6 @@ struct rcar_sysc_info {
 
 extern const struct rcar_sysc_info r8a7745_sysc_info;
 extern const struct rcar_sysc_info r8a7743_sysc_info;
+extern const struct rcar_sysc_info r8a7744_sysc_info;
 extern const struct rcar_sysc_info r8a77470_sysc_info;
 #endif /* __SOC_RENESAS_RCAR_SYSC_H__ */
\ No newline at end of file
-- 
2.7.4

